### DISCUSSION TITLE
Integration hooks for middleware

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#692] [Proposal]
### Body

# Summary

The ability for an integration to define middleware so that users do not need to go through the boilerplate of creating a middleware file and importing the functionality.

# Background & Motivation

When we built the middleware API we anticipated a need for integrations to hook into it, but we decided not to make that a goal for v1 due to the complexity of ordering. Yet the need keeps coming up for the same reasons as before, integrations want to transparently add middleware.

# Goals

- The ability to get middleware into an Astro project simply by adding an integration as normal.
- Allow an integration to place itself either at the front or end of the stack in terms of ordering.

# Example

```js
export default function() {
  return {
    name: 'astro-auth-middleware',
    hooks: {
      'astro:config:setup': ({ addMiddleware }) {
        addMiddleware({
          entrypoint: 'astro-auth-middleware/middleware',
          enforce: 'pre' // (or 'post')
        });
      }
    } 
  }
}
```

**Comment by alexanderniebuhr** [comment]
I'm curious: does the middleware need to strictly adhere to the Astro Middleware format? Or could it also take the form of a Cloudflare or Vercel Edge Middleware?

**Reply by ematipico** [reply]
Yes it needs to adhere to Astro middleware, because that's the expectation of the user, and that's how we chain it to other user middleware functions.

Middlewares of other platforms such as vercel and cloudflare have different signatures, functionalities and infrastructure so they need special treatment.

**Comment by ematipico** [comment]
What's the expected order of these functions when chaining them with the existing user middleware functions?

**Reply by matthewp** [reply]
With `pre` it would go before the user's middleware function. With `post` it would go last (right before the page/endpoint is called). Can bikeshed on the names and API a bit, but that's the general idea.

**Comment by matthewp** [comment]
Accepted for stage 2: https://github.com/withastro/roadmap/issues/709
